File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter17/FingerPaint/FingerPaint/MainPage.Share.cpp
#include "pch.h"
#include "MainPage.xaml.h"

using namespace FingerPaint;

using namespace Concurrency;
using namespace Platform;
using namespace Windows::ApplicationModel::DataTransfer;
using namespace Windows::Foundation;
using namespace Windows::Graphics::Imaging;
using namespace Windows::Storage::Streams;
using namespace Windows::System;
using namespace Windows::UI::Core;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Media::Imaging;


void MainPage::InitializeSharing()
{
    // Initialize the Past button and provide for updates
    CheckForPasteEnable();
    Clipboard::ContentChanged += ref new EventHandler<Object^>(this, &MainPage::OnClipboardContentChanged);

    // Watch for accelerator keys for Copy and Paste
    Window::Current->CoreWindow->Dispatcher->AcceleratorKeyActivated +=
        ref new TypedEventHandler<CoreDispatcher^, AcceleratorKeyEventArgs^>(this, &MainPage::OnAcceleratorKeyActivated);

    // Hook into the Share pane for providing data
    DataTransferManager::GetForCurrentView()->DataRequested +=
        ref new TypedEventHandler<DataTransferManager^, DataRequestedEventArgs^>(this, &MainPage::OnDataTransferDataRequested);
}

void MainPage::OnDataTransferDataRequested(DataTransferManager^ sender, DataRequestedEventArgs^ args)
{
    DataRequestDeferral^ deferral = args->Request->GetDeferral();

    // Get a stream reference and hand it over
    task<RandomAccessStreamReference^> getBitmapTask = create_task(GetBitmapStream(bitmap));
    getBitmapTask.then([args, deferral] (RandomAccessStreamReference^ streamRef)
    {
        args->Request->Data->SetBitmap(streamRef);
        args->Request->Data->Properties->Title = "Finger Paint";
        args->Request->Data->Properties->Description = "Share this painting with another app";

        deferral->Complete();
    });
}

void MainPage::OnClipboardContentChanged(Object^ sender, Object^ args)
{
    CheckForPasteEnable();
}

void MainPage::CheckForPasteEnable()
{
    pasteAppBarButton->IsEnabled = CheckClipboardForBitmap();
}

bool MainPage::CheckClipboardForBitmap()
{
    DataPackageView^ dataView = Clipboard::GetContent();
    return dataView->Contains(StandardDataFormats::Bitmap);
}

void MainPage::OnCopyAppBarButtonClick(Object^ sender, RoutedEventArgs^ args)
{
    task<RandomAccessStreamReference^> getBitmapTask = create_task(GetBitmapStream(bitmap));
    getBitmapTask.then([this] (RandomAccessStreamReference^ streamRef)
    {
        DataPackage^ dataPackage = ref new DataPackage();
        dataPackage->RequestedOperation = DataPackageOperation::Move;
        dataPackage->SetBitmap(streamRef);
        Clipboard::SetContent(dataPackage);
        this->BottomAppBar->IsOpen = false;
    });
}

task<RandomAccessStreamReference^> MainPage::GetBitmapStream(WriteableBitmap^ bitmap)
{
    // Create a BitmapEncoder associated with a memory stream
    InMemoryRandomAccessStream^ memoryStream = ref new InMemoryRandomAccessStream();
    task<BitmapEncoder^> createEncoderTask = create_task(BitmapEncoder::CreateAsync(BitmapEncoder::PngEncoderId,
                                                                                    memoryStream));
    return createEncoderTask.then([this, bitmap, memoryStream] (BitmapEncoder^ encoder)
    {
        // Create a pixels array for this bitmap
        int length = 4 * bitmap->PixelWidth * bitmap->PixelHeight;
        Array<byte>^ arrPixels = ref new Array<byte>(pixels, length);

        // Set the pixels into that encoder
        encoder->SetPixelData(BitmapPixelFormat::Bgra8,
                              BitmapAlphaMode::Premultiplied,
                              (unsigned int)bitmap->PixelWidth,
                              (unsigned int)bitmap->PixelHeight,
                              96, 96, arrPixels);

        task<void> flushEncoderTask = create_task(encoder->FlushAsync());
        return flushEncoderTask.then([memoryStream] ()
        {
            return RandomAccessStreamReference::CreateFromStream(memoryStream);
        });
    });
}

void MainPage::OnPasteAppBarButtonClick(Platform::Object^ sender, RoutedEventArgs^ args)
{
    // Temporarily disable the Paste button
    pasteAppBarButton->IsEnabled = false;

    // Get the Clipboard contents and check for a bitmap
    DataPackageView^ dataView = Clipboard::GetContent();

    if (dataView->Contains(StandardDataFormats::Bitmap))
    {
        // Get the stream reference and a stream
        task<RandomAccessStreamReference^> getBitmapTask = create_task(dataView->GetBitmapAsync());
        getBitmapTask.then([this] (RandomAccessStreamReference^ streamRef)
        {
            task<IRandomAccessStreamWithContentType^> openReadTask = create_task(streamRef->OpenReadAsync());
            openReadTask.then([this] (IRandomAccessStreamWithContentType^ stream)
            {
                task<BitmapDecoder^> createDecoderTask = create_task(BitmapDecoder::CreateAsync(stream));
                createDecoderTask.then([this] (BitmapDecoder^ decoder)
                {
                    task<BitmapFrame^> getFrameTask = create_task(decoder->GetFrameAsync(0));
                    getFrameTask.then([this] (BitmapFrame^ bitmapFrame)
                    {
                        // Save information for creating WriteableBitmap
                        pastedPixelWidth = (int)bitmapFrame->PixelWidth;
                        pastedPixelHeight = (int)bitmapFrame->PixelHeight;

                        task<PixelDataProvider^> getDataTask = 
                            create_task(bitmapFrame->GetPixelDataAsync(BitmapPixelFormat::Bgra8,
                                                                       BitmapAlphaMode::Premultiplied,
                                                                       ref new BitmapTransform(),
                                                                       ExifOrientationMode::RespectExifOrientation,
                                                                       ColorManagementMode::ColorManageToSRgb));

                        getDataTask.then([this] (PixelDataProvider^ pixelProvider)
                        {
                            // Save information sufficient for creating WriteableBitmap
                            pastedPixels = pixelProvider->DetachPixelData();
                            // Check if it's OK to replace the current painting
                            task<void> checkOkToTrashTask = create_task(CheckIfOkToTrashFile(&MainPage::FinishPasteBitmapAndPixelArray));
                            checkOkToTrashTask.then([this] ()
                            {
                                // Re-enable the button and close th app bar
                                pasteAppBarButton->IsEnabled = true;
                                this->BottomAppBar->IsOpen = false;
                            }, task_continuation_context::use_current());
                        });
                    });
                });
            });
        });
    }
}

Concurrency::task<void> MainPage::FinishPasteBitmapAndPixelArray()
{
    bitmap = ref new WriteableBitmap(pastedPixelWidth, pastedPixelHeight);

    InitializeBitmap();

    // Transfer pixels to bitmap
    int length = 4 * pastedPixelWidth * pastedPixelHeight;

    for (int i = 0; i < length; i++)
        pixels[i] = pastedPixels[i];

    pastedPixels = nullptr;

    // Set AppSettings properties for new image
    appSettings->LoadedFilePath = nullptr;
    appSettings->LoadedFilename = nullptr;
    appSettings->IsImageModified = false;

    return task<void>([] () {});
}

void MainPage::OnAcceleratorKeyActivated(CoreDispatcher^ sender, AcceleratorKeyEventArgs^ args)
{
    if ((args->EventType == CoreAcceleratorKeyEventType::SystemKeyDown ||
         args->EventType == CoreAcceleratorKeyEventType::KeyDown) &&
        (args->VirtualKey == VirtualKey::C || args->VirtualKey == VirtualKey::V))
    {
        CoreWindow^ window = Window::Current->CoreWindow;
        CoreVirtualKeyStates down = CoreVirtualKeyStates::Down;

        // Only want case where Ctrl is down
        if ((window->GetKeyState(VirtualKey::Shift) & down) == down ||
            (window->GetKeyState(VirtualKey::Control) & down) != down ||
            (window->GetKeyState(VirtualKey::Menu) & down) == down)
        {
            return;
        }

        if (args->VirtualKey == VirtualKey::C)
        {
            OnCopyAppBarButtonClick(nullptr, nullptr);
        }
        else if (args->VirtualKey == VirtualKey::V)
        {
            OnPasteAppBarButtonClick(pasteAppBarButton, nullptr);
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter17/FingerPaint/FingerPaint/MainPage.Share.cs
using System;
using System.IO;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Windows.ApplicationModel.DataTransfer;
using Windows.Graphics.Imaging;
using Windows.Storage.Streams;
using Windows.System;
using Windows.UI.Core;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media.Imaging;

namespace FingerPaint
{
    public sealed partial class MainPage : Page
    {
        int pastedPixelWidth, pastedPixelHeight;
        byte[] pastedPixels;

        void InitializeSharing()
        {
            // Initialize the Paste button and provide for updates
            CheckForPasteEnable();
            Clipboard.ContentChanged += OnClipboardContentChanged;

            // Watch for accelerator keys for Copy and Paste
            Window.Current.CoreWindow.Dispatcher.AcceleratorKeyActivated += 
                                                                    OnAcceleratorKeyActivated;

            // Hook into the Share pane for providing data
            DataTransferManager.GetForCurrentView().DataRequested += OnDataTransferDataRequested;
        }

        async void OnDataTransferDataRequested(DataTransferManager sender, DataRequestedEventArgs args)
        {
            DataRequestDeferral deferral = args.Request.GetDeferral();

            // Get a stream reference and hand it over
            RandomAccessStreamReference reference = await GetBitmapStream(bitmap);
            args.Request.Data.SetBitmap(reference);
            args.Request.Data.Properties.Title = "Finger Paint";
            args.Request.Data.Properties.Description = "Share this painting with another app";

            deferral.Complete();
        }

        void OnClipboardContentChanged(object sender, object args)
        {
            CheckForPasteEnable();
        }

        void CheckForPasteEnable()
        {
            pasteAppBarButton.IsEnabled = CheckClipboardForBitmap();
        }

        bool CheckClipboardForBitmap()
        {
            DataPackageView dataView = Clipboard.GetContent();
            return dataView.Contains(StandardDataFormats.Bitmap);
        }

        async void OnCopyAppBarButtonClick(object sender, RoutedEventArgs args)
        {
            DataPackage dataPackage = new DataPackage
            {
                 RequestedOperation = DataPackageOperation.Move,
            };
            dataPackage.SetBitmap(await GetBitmapStream(bitmap));

            Clipboard.SetContent(dataPackage);
            this.BottomAppBar.IsOpen = false;
        }

        async Task<RandomAccessStreamReference> GetBitmapStream(WriteableBitmap bitmap)
        {
            // Get a pixels array for this bitmap
            byte[] pixels = new byte[4 * bitmap.PixelWidth * bitmap.PixelHeight];
            Stream stream = bitmap.PixelBuffer.AsStream();
            await stream.ReadAsync(pixels, 0, pixels.Length);

            // Create a BitmapEncoder associated with a memory stream
            InMemoryRandomAccessStream memoryStream = new InMemoryRandomAccessStream();
            BitmapEncoder encoder = await BitmapEncoder.CreateAsync(BitmapEncoder.PngEncoderId, 
                                                                    memoryStream);

            // Set the pixels into that encoder
            encoder.SetPixelData(BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied,
                                 (uint)bitmap.PixelWidth, (uint)bitmap.PixelHeight, 96, 96, pixels);
            await encoder.FlushAsync();

            // Return a RandomAccessStreamReference
            return RandomAccessStreamReference.CreateFromStream(memoryStream);
        }

        async void OnPasteAppBarButtonClick(object sender, RoutedEventArgs args)
        {
            // Temporarily disable the Paste button
            Button button = sender as Button;
            button.IsEnabled = false;

            // Get the Clipboard contents and check for a bitmap
            DataPackageView dataView = Clipboard.GetContent();

            if (dataView.Contains(StandardDataFormats.Bitmap))
            {
                // Get the stream reference and a stream
                RandomAccessStreamReference streamRef = await dataView.GetBitmapAsync();
                IRandomAccessStreamWithContentType stream = await streamRef.OpenReadAsync();

                // Create a BitmapDecoder for reading the bitmap
                BitmapDecoder decoder = await BitmapDecoder.CreateAsync(stream);
                BitmapFrame bitmapFrame = await decoder.GetFrameAsync(0);
                PixelDataProvider pixelProvider = 
                    await bitmapFrame.GetPixelDataAsync(BitmapPixelFormat.Bgra8, 
                                                        BitmapAlphaMode.Premultiplied, 
                                                        new BitmapTransform(),
                                                        ExifOrientationMode.RespectExifOrientation, 
                                                        ColorManagementMode.ColorManageToSRgb);

                // Save information sufficient for creating WriteableBitmap
                pastedPixelWidth = (int)bitmapFrame.PixelWidth;
                pastedPixelHeight = (int)bitmapFrame.PixelHeight;
                pastedPixels = pixelProvider.DetachPixelData();

                // Check if it's OK to replace the current painting
                await CheckIfOkToTrashFile(FinishPasteBitmapAndPixelArray);
            }

            // Re-enable the button and close the app bar
            button.IsEnabled = true;
            this.BottomAppBar.IsOpen = false;
        }

        async Task FinishPasteBitmapAndPixelArray()
        {
            bitmap = new WriteableBitmap(pastedPixelWidth, pastedPixelHeight);
            pixels = pastedPixels;

            // Transfer pixels to bitmap, among other chores
            await InitializeBitmap();

            // Set AppSettings properties for new image
            appSettings.LoadedFilePath = null;
            appSettings.LoadedFilename = null;
            appSettings.IsImageModified = false;
        }

        void OnAcceleratorKeyActivated(CoreDispatcher sender, AcceleratorKeyEventArgs args)
        {
            if ((args.EventType == CoreAcceleratorKeyEventType.SystemKeyDown ||
                 args.EventType == CoreAcceleratorKeyEventType.KeyDown) &&
                (args.VirtualKey == VirtualKey.C || args.VirtualKey == VirtualKey.V))
            {
                CoreWindow window = Window.Current.CoreWindow;
                CoreVirtualKeyStates down = CoreVirtualKeyStates.Down;

                // Only want case where Ctrl is down
                if ((window.GetKeyState(VirtualKey.Shift) & down) == down ||
                    (window.GetKeyState(VirtualKey.Control) & down) != down ||
                    (window.GetKeyState(VirtualKey.Menu) & down) == down)
                {
                    return;
                }

                if (args.VirtualKey == VirtualKey.C)
                {
                    OnCopyAppBarButtonClick(null, null);   
                }
                else if (args.VirtualKey == VirtualKey.V)
                {
                    OnPasteAppBarButtonClick(pasteAppBarButton, null);
                }
            }
        }
    }
}
